tool.multipath: warn when user_friendly_names is enabled

aesteve-rh opened the pull request:
This PR updates the isconfigured check for the multipath in the vdsm-tool,
so that running:

    # vdsm-tool is-configured --module multipath

Will print a WARN message if there is any device or section in the multipath daemon configuration that enables 'user_friendly_names', as in:

```
WARNING: Invalid configuration: 'user_friendly_names' is enabled in multipath configuration:
  defaults {
    key value
    user_friendly_names yes
  }
  device {
    user_friendly_names yes
    key value
  }
This configuration is not supported and may lead to storage domain corruption.
```

aesteve-rh commented:
This is probably worth a test. I have to check how the tool modules are tested, I took a brief look at it last week.

nirs commented on lib/vdsm/tool/configurators/multipath.py:
```diff
@@ -327,44 +327,60 @@ def configure():
+        bool: check has passed.
+    """
+    first = mpathconf[0]
+    second = mpathconf[1]
```
Letting IndexError propagate here is not great. We don't want that the caller will have to know that we raise IndexError, and this can hide real error in other part of the function.

The error should be handled here, and we need to log a good error and return the right value to the caller.

aesteve-rh replied:
Uhm, afterwards I change it for `readline()` that avoids the exception. I'll reorder the patches so that it happens first, and we don't need to worry about the exception at all.

nirs commented on lib/vdsm/tool/configurators/multipath.py:
```diff
@@ -327,44 +327,60 @@ def configure():
 
 
-def isconfigured():
+def _check_conf_rev(mpathconf):
```
We need more clear names like _check_conf_revision()

nirs commented on lib/vdsm/tool/configurators/multipath.py:
```diff
@@ -327,44 +327,60 @@ def configure():
-            second = mpathconf[1]
+            conf_check = _check_conf_rev(mpathconf)
+            return YES if conf_check else NO
         except IndexError:
```
Do we want to fail on the first check or do all the checks and report all errors?
Reporting all errors at once may be more user friendly but harder to get right.

If we want to check all, and keep the results, lets use more clear names:

    revision_ok = _check_revision(mpathconf)
    user_friendly_names_ok = _check_user_friendly_names(mpathconf)

    report errors...

nirs commented on lib/vdsm/tool/configurators/multipath.py:
```diff
@@ -327,22 +327,20 @@ def configure():
-        mpathconf (List[str]): Lines in the multipath daemon
-        configuration file.
+        first (str): First line in the multipath daemon configuration file.
+        second (str): Second line in the multipath daemon configuration file.
```
Passing here the first lines of the file is not an improvement. Please keep the code as is, passing the contents of the file as lines.

aesteve-rh replied:
OK. But it can be a list with two elements so that we can do:

     first, second = mpathconf

and not raise (with `readline()`).

nirs replied:
But mpathconf should be the entire file. The caller should not do the work for the function.

nirs commented on lib/vdsm/tool/configurators/multipath.py:
```diff
@@ -327,22 +327,20 @@ def configure():
         bool: check has passed.
     """
-    first = mpathconf[0]
-    second = mpathconf[1]
```
Please keep this, and handle the IndexError here.

nirs commented on lib/vdsm/tool/configurators/multipath.py:
```diff
@@ -372,15 +370,13 @@ def isconfigured():
     """
-
-    if os.path.exists(_CONF_FILE):
+    if os.path.exists(_CONF_FILE) and os.stat(_CONF_FILE).st_size > 0:
```
Empty /etc/multipath.conf is not a valid configuration and we should not keep special code to handle it.

The file cannot be empty, and must contain the "RHEV REVISION X.Y" and "RHEV PRIVATE"
lines. If not the configuration is invalid and vdsm should not start!

aesteve-rh replied:
And it does. If the file exists **and has contents**, we verify contents, otherwise, we fail.

The only thing that changed is the "and" part.

nirs commented on lib/vdsm/tool/configurators/multipath.py:
```diff
@@ -372,15 +370,13 @@ def isconfigured():
+                             for x in (f.readline() for _ in range(2))]
+
+        conf_check = _check_conf_rev(first, second)
+        return YES if conf_check else NO
```
This code should be kept as is, just remove the error handler since _check_conf_revision() should handle it.

    if os.path.exists(_CONF_FILE):
        with open(_CONF_FILE) as f:
            mpathconf = [x.strip("\n") for x in f.readlines()]
            revision_ok = _check_revision(first, second)
            ...

nirs commented:
> This check may be run when vdsm starts (in a hook?).

All configurations run on every vdsm starts via "vdsm-tool is-configured".
If the configurator returns NO, vdsm will fail to start.

nirs commented on lib/vdsm/storage/mpathconf.py:
```diff
@@ -428,3 +429,26 @@ def read_metadata():
+    # - Between the section string and the left bracket, a single space is
+    #   allowed as maximum.
+    # - Between keys and values at least one space is required.
+    pattern = r'\s+(\S+) +{[^{]+user_friendly_names +"?yes"?\s+}'
```
Few issues:
- The section name always starts with leading whitespace - how does it work for

      defaults {

      }

- This assumes that "user_friendly_names yes" is the last line before the "}"

This would be more clear if we use named regex and verbose option:

aesteve-rh replied:
I think I tried key-values after the "user_friendly_names yes" and worked, but you are right that it should only accept "blank" characters, so maybe I didn't test it right.

I am currently trying to clear a few of the assumptions to rework the regex, but I'll take these comments into consideration next time I push. Possibly also add some simple tests to increase the coverage.

I'll draft the PR for now, until I figure out the small opens.

aesteve-rh commented on lib/vdsm/storage/multipath.py:
```diff
@@ -209,6 +209,18 @@ def is_ready():
     return "busy: false" in status
 
 
+def get_local_config():
```
`show_config_local` sound good :+1:

nirs commented:
> Will print a WARN message if there is any device or section in the multipath daemon configuration that enables 'user_friendly_names', as in:
> 
> ```
>     WARNING: invalid configuration in /etc/multipath.conf:
>     device_name -> user_friendly_names:  yes
>     This configuration is not supported and may lead storage
>     domain corruption.
> ```

Issues:

- The configuration is likely not in /etc/multpath.conf buf from a drop-in file
  in /etc/multipath/conf.d/*.conf, and multipathd does not report the source of
  the configuration value.
- device_name cannot work because the configuration can come from defaults,
  overdides, devices, or multpaths. Devices and multipaths do not have a name
  but we can show the properties of the device or multipath entry.

We can use this:

    user_friendly_names is enabled for 'defaults' key=value ...
    user_friendly_names is enabled for 'overrides' key=value ...
    user_friendly_names is enabled for 'device' key=value ...
    user_friendly_names is enabled for 'multiapth' key=value ...
   
So the check function can collect the name and the attributes of any
unsupported block of confiugration.

nirs commented on lib/vdsm/storage/mpathconf.py:
```diff
@@ -428,3 +432,95 @@ def read_metadata():
+    return None, True
+
+
+def _parse_conf(mpath_out, parsed_conf, depth=0):
```
We can simplify this by accepting an iterator instead of list of lines.

For example the caller can use:

     import io
     it = io.StringIO(out)
     tree = _parse_conf(it)

The advantage of the iterator is that we don't need to manage the depth or pass the list to append value into. Instead we can just return the parsed value.

aesteve-rh replied:
Nice, I am not very familiar with the io functions, so I didn't know this could be done so easily.

This will make parsing a lot easier, thanks!

nirs commented on lib/vdsm/storage/mpathconf.py:
```diff
@@ -428,3 +432,95 @@ def read_metadata():
+    i = 0
+    while i < len(mpath_out):
+        line = mpath_out[i].strip()
+        i += 1
```
This is good way to do this in C, but we can make this much simpler using
iteration. Assuming the caller sent us an iterator over the lines of the config:

    for line in mpath_out:
        process line...

aesteve-rh replied:
This does not work if you want to jump iterations. That's why I used index-based iteration.
But with the iterator suggested above I can change this easily.

And for C, at least in C++ we have range-based loops available since C++11 :)

nirs commented on lib/vdsm/storage/mpathconf.py:
```diff
@@ -428,3 +432,95 @@ def read_metadata():
+            name, _ = line.rsplit(" ", 1)
+            nested_section = []
+            i += _parse_conf(mpath_out[i:], nested_section, depth + 1)
+            parsed_conf.append(ConfSection(name, nested_section))
```
If we return the parsed config, this becomes much simpler:

    res.append(name, _parse_conf(mpath_out))

aesteve-rh replied:
Again, byproduct of not using iterator and wanting to jump iterations. Will get fixed with StringIO.

nirs commented on lib/vdsm/storage/mpathconf.py:
```diff
@@ -428,3 +432,95 @@ def read_metadata():
+            parsed_conf.append(ConfSection(name, nested_section))
+        elif line == "}":
+            if depth > 0:
+                return i
```
It is better to break here and return at the bottom. What if the last } is missing?
The current code will return None.

nirs commented on lib/vdsm/storage/mpathconf.py:
```diff
@@ -428,3 +432,95 @@ def read_metadata():
+        else:
+            if not line:
+                # Discard empty lines
+                continue
```
We don't have empty lines in multipathd output, but it will be more robust to ignore such lines.

nirs commented on lib/vdsm/storage/mpathconf.py:
```diff
@@ -297,6 +299,8 @@
 
 Metadata = namedtuple("Metadata", "revision,private")
+ConfSection = namedtuple("ConfSection", "name,attrs")
+ConfPair = namedtuple("ConfPair", "key,value")
```
Since this is the mpathconf module, why not use Section and Pair?

nirs commented on tests/storage/mpathconf_test.py:
```diff
@@ -49,6 +52,32 @@
+        prio_args ""
+        user_friendly_names $defaults_value
+        find_multipaths_timeout -10
+        enable_foreign "NONE"
```
I already commented about this - you test here only value in the middle, but we need to test:

- value is first
- value is last
- value is not first or last

This is true for all sections. So using one test config is not helpful.

For testing parsing, we don't care about actual keys - it is enough to use
only few keys, possibly always:

    section {
        key1 "value1"
        key2 "value2"
        key3 "value3"
    }

It would be nice to test sections with no key, 1 key, and multiple keys.

The parser does not care abut the actual names of the sections, it depends
only on the generic format (key, value). So we can test the parser with:

    section1 {
        key1 "value1"
        key2 "value2"
        key3 "value3"
    }
    section2 {
        key1 "value1"
    }
    section3 {
    }
    section4 {
        entry {
            key1 "value1"
            key2 "value2"
            key3 "value3"
        }
        entry {
            key1 "value1"
        }
        entry {
        }
    }

This will help to understand the parser and make the parser test very clear.
For example we can test:

    assert _parse_conf(it) == [
         Section(
             "section1", [
                 Pair("key1", "value1"),
                 ...
             ]
         ),
         ...
    ]

aesteve-rh replied:
I didn't ignore the point. The cases of:
- value is first
- value is last
- value is not first or last
I addressed them by putting the user_friendly_names pair in different locations in different sections.

But nevermind that. I'll separate the config example in different blocks and tests, with generic names.

nirs commented on tests/storage/mpathconf_test.py:
```diff
@@ -49,6 +52,32 @@
+    }
+    multipath {
+        wwid "wwid2"
+    }
```
You need to test also "devices" section.

aesteve-rh replied:
I'll use generic names instead.

nirs commented on tests/storage/mpathconf_test.py:
```diff
@@ -172,3 +201,46 @@ def test_mpath_empty_conf(fake_mpath_conf):
+    (["no", "yes", "no"], "overrides"),
+    (['"no"', '"yes"', '"no"'], "overrides"),
+    (["no", "no", "yes"], "multipath"),
+    (['"no"', '"no"', '"yes"'], "multipath"),
```
This is very hard to understand. what is the first "yes", "no", "yes"?

Don't try to create the shortest possible code. Try to create the easier to 
understand code. Test must the simple as possible to ensure they have no bugs.

aesteve-rh replied:
Fair point. Although I don't think it is that difficult to understand, basically the first line of the function is telling you what each value means. And I've seen more complex uses of the parameters in tests. But ok, I'll separate the parser tests into simpler examples, and the checker itself into a separate set of tests and see how it looks.

aesteve-rh replied:
So finally I kept a test with parameters, but simplified the way the test is parametrized so that it is more readable.

And complemented the test with specific tests for the position of the user_friendly_names attribute in a section.

IMO this is nicer than what I did in the previous version of my parametrized code, and nicer than copy&pasting code multiple times for different section. Let me know what you think.

nirs commented on tests/storage/mpathconf_test.py:
```diff
@@ -172,3 +201,46 @@ def test_mpath_empty_conf(fake_mpath_conf):
+        mpathconf.check_local_config()
+    assert section == e.value.context["section"].name
+    invalid_pair = str(mpathconf.ConfPair('user_friendly_names', 'yes'))
+    assert invalid_pair in str(e.value)
```
I think we need simple test of the parser before we test the code validating the parsed tree.

nirs commented on tests/storage/mpathconf_test.py:
```diff
@@ -172,3 +201,46 @@ def test_mpath_empty_conf(fake_mpath_conf):
+    assert invalid_pair in str(e.value)
+
+
+@pytest.mark.parametrize("value", ["no", '"no"'])
```
Using parameters for testing unquoting is nice and simple, but this can be tested in in the parser test. Once we tested that quoted values are unquoted, we can test only quoted values when checking detection of bad options.

nirs commented on lib/vdsm/tool/configurators/multipath.py:
```diff
@@ -109,3 +106,27 @@ def isconfigured():
+            sys.stdout.write(f"{line_indent}{attr.key} {attr.value}\n")
+        sys.stdout.write("  }\n")
+        sys.stdout.write("This configuration is not supported "
+                         "and may lead storage domain corruption.\n")
```
Please share the output printed here in the commit message.

nirs commented on lib/vdsm/tool/configurators/multipath.py:
```diff
@@ -109,3 +106,27 @@ def isconfigured():
+        for attr in ufn_enabled_section.attrs:
+            line_indent = "    "
+            if attr.key == "user_friendly_names":
+                line_indent = "->" + line_indent
```
I don't think we need this since we start the warning with "user_friendly_names is enabled". Showing the config as is simpler and better.

nirs commented on lib/vdsm/tool/configurators/multipath.py:
```diff
@@ -109,3 +106,27 @@ def isconfigured():
+        sys.stdout.write("  }\n")
+        sys.stdout.write("This configuration is not supported "
+                         "and may lead storage domain corruption.\n")
+        return NO
```
Returning NO means that vdsm will not start on a system with unsupported configuration. This may break existing setups so we may want to just warn about the issue without failing vdsm-tool.

We can start by warning, and later convert the warning to hard error if needed.

aesteve-rh replied:
The whole point of the bug fix is making it fail to avoid wrong configurations, no? The issue may get unnoticed if the upgrade tool logs are not checked.

aesteve-rh replied:
Ok. Let's go for it then, and start with a warning. My main concern is in that the warning can be easily missed, but I understand that breaking after an upgrade is not something desirable for users either.

aesteve-rh commented on lib/vdsm/storage/mpathconf.py:
```diff
@@ -428,3 +430,33 @@ def read_metadata():
     return Metadata(REVISION_MISSING, private)
+
+
+def _parse_conf(mpath_out):
```
I'll append `_it` to the name.

aesteve-rh commented on lib/vdsm/storage/mpathconf.py:
```diff
@@ -432,6 +434,62 @@ def read_metadata():
+                f"{name}.{attr.name}", attr.attrs)
+            if not check_ok:
+                return attr, check_ok
+    return None, True
```
As a possibility, if all the children must have the same type, then we can use `all` function to avoid checking only the first child. It is less performant but here is not important, and with the data size that the multipath configuration outputs should not matter anyway.

aesteve-rh replied:
Then I can use `all` here, that will fix checking only the first child and ensure consistency in the sections.

aesteve-rh commented on tests/storage/mpathconf_test.py:
```diff
@@ -51,6 +53,46 @@
 }
 """
 
+FAKE_MPATHD_OUTPUT = Template("""\
```
There is no verification intention in the pairs here. The intent is just having a realistic output, as a hint to other maintainers. To me, it is nice having a real feel of the data that we will end up receiving.

But if you see not value in that I can standardize them with "key value" pairs.

aesteve-rh commented on lib/vdsm/storage/mpathconf.py:
```diff
@@ -432,6 +434,62 @@ def read_metadata():
+
+    Raises:
+        exception.InvalidConfiguration: if any section have
+            'user_friendly_names' enabled. Contains the offending section name.
```
Updated also to contain multiple sections.

aesteve-rh commented on tests/storage/mpathconf_test.py:
```diff
@@ -217,3 +259,79 @@ def test_mpathd_parser():
+        monkeypatch.setattr(
+            mpathconf.multipath, 'show_config_local', mockreturn)
+
+    return _factory
```
I have finally removed the `fixture`.

aesteve-rh commented:
I have addressed all comments. I left a couple threads open for you to read (@nirs), but still I have followed them too.

Hopefully this will be closer to a mergeable version.

nirs commented on tests/storage/mpathconf_test.py:
```diff
@@ -224,3 +224,157 @@ def test_mpathd_parser():
+    fake_conf = [
+        Section('default', [
+            Pair('key', 'value'),
+            Pair('user_friendly_names', 'yes'),
```
I think this covers the last pair case...

nirs commented on tests/storage/mpathconf_test.py:
```diff
@@ -224,3 +224,157 @@ def test_mpathd_parser():
+        ]),
+        Section('devices', [
+            Section('device', [
+                Pair('user_friendly_names', 'yes'),
```
and this cover the first pair case. If we add another device or multipath to the test data using middle pair, we cover the case.

nirs commented on tests/storage/mpathconf_test.py:
```diff
@@ -224,3 +224,157 @@ def test_mpathd_parser():
+    ]
+
+
+def test_mpathd_ufn_first_pair():
```
Already covered int the previous tests.

nirs commented on tests/storage/mpathconf_test.py:
```diff
@@ -224,3 +224,157 @@ def test_mpathd_parser():
+    ]
+
+
+def test_mpathd_ufn_mid_pair():
```
Not covered in the previous tests but can be easily added to the multiple test.

nirs commented on tests/storage/mpathconf_test.py:
```diff
@@ -224,3 +224,157 @@ def test_mpathd_parser():
+    ]
+
+
+def test_mpathd_ufn_last_pair():
```
Already covered by previous tests.

nirs commented on tests/storage/mpathconf_test.py:
```diff
@@ -224,3 +224,157 @@ def test_mpathd_parser():
+    fake_conf = [
+        Section('section1', [
+            Pair('user_friendly_names', 'yes'),
+        ])
```
This can also be added to the previous tests, maybe the multiple test?

nirs commented on tests/storage/mpathconf_test.py:
```diff
@@ -224,3 +224,157 @@ def test_mpathd_parser():
+
+def test_mpathd_ufn_enabled_single():
+    fake_conf = [
+        Section('default', [
```
Just for consistency, this is called "defaults". If you send another version it would be nice to rename in all tests.

nirs commented on tests/storage/mpathconf_test.py:
```diff
@@ -224,3 +224,157 @@ def test_mpathd_parser():
+        Section('section1', [])
+    ]
+    issues = mpathconf._check_conf(fake_conf)
+    assert not issues
```
Testing empty list will be nice for completeness.

nirs commented on lib/vdsm/tool/configurators/multipath.py:
```diff
@@ -109,3 +106,25 @@ def isconfigured():
+                sys.stdout.write(f"    {attr.key} {attr.value}\n")
+            sys.stdout.write("  }\n")
+            sys.stdout.write("This configuration is not supported "
+                             "and may lead to storage domain corruption.\n")
```
It would be better to use a specific error and format the error message in the exception class __str__(). The we will simplify do there:

    try:
        check config...
    except InvalidMultipathConfiguration as e:
        sys.stdout.write("WARNING: %s", e)

aesteve-rh replied:
I think is good as it is. The exception is a generic container with a bit of context.

`sections` contain the raw contents of the configuration structure, so it will get printed as:
```
Invalid configuration value: 'user_friendly_names' is enabled in multipath configuration {'sections': [Section(name='defaults', children=[Pair(key='key', value='value'), Pair(key='user_friendly_names', value='yes')]), Section(name='device', children=[Pair(key='user_friendly_names', value='yes'), Pair(key='key', value='value')])]}
```

Which is horrible for users. This structure needs to get transformed into a readable format, but the exception class is agnostic to whatever the context dictionary may contain. Then, we either transform it before in the mpathconf module or after, in the configurator. IMO, the configurator is a good place to do it, is where we have full control over what we want to show.

In here I just need to get the first and last lines out of the loop, and it looks good. I indented it wrongly in some update.

nirs replied:
The exception __str__ can return a string with multiple lines - the same thing you output here. But it is true that it will be harder to control how the output look in the tool.

If you want to format the message here, use more specific exception that has a public
"issues" or "sections" attribute instead of abusing the context dict.

aesteve-rh replied:
Something like this?

```python
class InvalidMPathConfiguration(InvalidConfiguration):
    code = 103
    message = "Invalid configuration value"

    def __init__(self, reason, sections):
        super().__init__(reason)
        self.sections = sections
    
    def __str__(self):
        error_str = f"{self.msg}: {self.context['reason']}:\n"
        for section in self.sections:
            error_str += f"  {section.name} {{\n"
            for attr in section.children:
                error_str += f"    {attr.key} {attr.value}\n"
            error_str += "  }\n"
        error_str += "This configuration is not supported " \
                     "and may lead to storage domain corruption.\n"
        return error_str
```

nirs replied:
I would use io.StringIO to format, but this is the idea. If you do the formatting
in `__str__` you don't need to provide a public `sections` attribute. If you want to format the error in the tool, using a public attribute is better.

aesteve-rh replied:
StringIO sounds good. I will probably format the error in the _ _str_ _. As we are introducing a new exception class anyway, my argument above about keeping it agnostic to context is not important.

nirs replied:
At the end what matters is what works best for our needs. The general approach in vdsm to to format errors in a useful way in the exception class. However we sue this only for logging when we care less about the formatting and more about the content. If formatting the error in the tool is easier do it in the tool.

Another option is to avoid the exception and instead return the result of the check to the tool, and format the warning in the tool. Even if we want to fail, we can do this
in the tool instead of the mpathconf module.

nirs commented on lib/vdsm/tool/configurators/multipath.py:
```diff
@@ -109,3 +106,25 @@ def isconfigured():
+    try:
+        mpathconf.check_local_config()
+    except exception.InvalidConfiguration as e:
+        ufn_enabled_sections = e.context["sections"]
```
Using the context works, but it will be better to handle error formatting in the exception class.

nirs commented on lib/vdsm/common/exception.py:
```diff
@@ -505,6 +505,18 @@ class MigrationOperationError(VdsmException):
+        self.sections = sections
+
+    def __str__(self):
+        return f"{self.msg}: {self.context['reason']}"
```
Why do we need to override `__str__`? The only reason I can see is include the `sections` in the log, but you don't include it.

The easiest way to handle this is to inherit the super class `__str__` and make sure the sections are included in the context, which is already part of the supervlass `__str__`.

A possible solution can be:

    class InvalidMPathConfiguration(ContextException):
        ...
        @property
        def sections(self):
            return self.context["sections"]

Now we have good `__str__` without writing any code, and we have a public property to access the sections that can be used by the tool to format a warning.

aesteve-rh replied:
The `__str__` method is overriden because the parent `__str__` returns a raw dictionary as in `{"reason": "this is the reason", "another_key": "another_value"}`, which is fine for the text of an exception but a bit ugly for the output of a command line tool.

Since I couldn't use the context dictionary from the tool where the exception is catched, I did override the string method.
I can expose the `reason` (and `context`) keys through properties as you suggest, but it is mostly a cosmetic discussion.

nirs commented on lib/vdsm/tool/configurators/multipath.py:
```diff
@@ -109,3 +106,24 @@ def isconfigured():
+        mpathconf.check_local_config()
+    except exception.InvalidMPathConfiguration as e:
+        # Just warn for now, do not fail the isconfigured check here.
+        sys.stdout.write(f"WARNING: {e}:\n")
```
Using the error here means the error `__str__` cannot include all the info about the error, which is bad if we want to log it.

In the context of the tool, you should use your own message instead of using the error string representation.

Looks like you make the work harder by raising an error from mpathconf module, instead of reporting the status and using it here.

if we use:

    result = mpathconf.check_local_config()

And result is a named tuple with:
- error
- issues

We should format here the error and the issues, and we don't need to introduce a new exception. This is how lvmfilter and lvmdevices module work.

aesteve-rh replied:
That is also how this was coded before, until I changed it following a comment. I don't mind removing the exception and bringing that back again.

nirs commented:
/ost

nirs commented:
OST failed when trying to start vdsmd service:

```
2022-07-25T13:31:46.609500+0000 ost-basic-suite-master-host-0 vdsmd_init_common.sh[40889]: vdsm: Running check_is_configured
2022-07-25T13:31:48.055712+0000 ost-basic-suite-master-host-0 vdsmd_init_common.sh[40924]: Traceback (most recent call last):
2022-07-25T13:31:48.055712+0000 ost-basic-suite-master-host-0 vdsmd_init_common.sh[40924]:   File "/usr/bin/vdsm-tool", line 209, in main
2022-07-25T13:31:48.055712+0000 ost-basic-suite-master-host-0 vdsmd_init_common.sh[40924]:     return tool_command[cmd]["command"](*args)
2022-07-25T13:31:48.055712+0000 ost-basic-suite-master-host-0 vdsmd_init_common.sh[40924]:   File "/usr/lib/python3.6/site-packages/vdsm/tool/__init__.py", line 40, in wrapper
2022-07-25T13:31:48.055712+0000 ost-basic-suite-master-host-0 vdsmd_init_common.sh[40924]:     func(*args, **kwargs)
2022-07-25T13:31:48.055712+0000 ost-basic-suite-master-host-0 vdsmd_init_common.sh[40924]:   File "/usr/lib/python3.6/site-packages/vdsm/tool/configurator.py", line 165, in isconfigured
2022-07-25T13:31:48.055712+0000 ost-basic-suite-master-host-0 vdsmd_init_common.sh[40924]:     m = [c.name for c in pargs.modules if _isconfigured(c) == configurators.NO]
2022-07-25T13:31:48.055712+0000 ost-basic-suite-master-host-0 vdsmd_init_common.sh[40924]:   File "/usr/lib/python3.6/site-packages/vdsm/tool/configurator.py", line 165, in <listcomp>
2022-07-25T13:31:48.055712+0000 ost-basic-suite-master-host-0 vdsmd_init_common.sh[40924]:     m = [c.name for c in pargs.modules if _isconfigured(c) == configurators.NO]
2022-07-25T13:31:48.055712+0000 ost-basic-suite-master-host-0 vdsmd_init_common.sh[40924]:   File "/usr/lib/python3.6/site-packages/vdsm/tool/configurator.py", line 103, in _isconfigured
2022-07-25T13:31:48.055712+0000 ost-basic-suite-master-host-0 vdsmd_init_common.sh[40924]:     return getattr(module, 'isconfigured', lambda: configurators.NO)()
2022-07-25T13:31:48.055712+0000 ost-basic-suite-master-host-0 vdsmd_init_common.sh[40924]:   File "/usr/lib/python3.6/site-packages/vdsm/tool/configurators/multipath.py", line 115, in isconfigured
2022-07-25T13:31:48.055712+0000 ost-basic-suite-master-host-0 vdsmd_init_common.sh[40924]:     check_result = mpathconf.check_local_config()
2022-07-25T13:31:48.055712+0000 ost-basic-suite-master-host-0 vdsmd_init_common.sh[40924]:   File "/usr/lib/python3.6/site-packages/vdsm/storage/mpathconf.py", line 447, in check_local_config
2022-07-25T13:31:48.055712+0000 ost-basic-suite-master-host-0 vdsmd_init_common.sh[40924]:     mpath_conf = _parse_conf(io.StringIO(multipath.show_config_local()))
2022-07-25T13:31:48.055712+0000 ost-basic-suite-master-host-0 vdsmd_init_common.sh[40924]:   File "/usr/lib/python3.6/site-packages/vdsm/storage/multipath.py", line 221, in show_config_local
2022-07-25T13:31:48.055712+0000 ost-basic-suite-master-host-0 vdsmd_init_common.sh[40924]:     [_MULTIPATHD.cmd, "show", "config", "local"]).decode('ascii')
2022-07-25T13:31:48.055712+0000 ost-basic-suite-master-host-0 vdsmd_init_common.sh[40924]:   File "/usr/lib/python3.6/site-packages/vdsm/common/commands.py", line 102, in run
2022-07-25T13:31:48.055712+0000 ost-basic-suite-master-host-0 vdsmd_init_common.sh[40924]:     raise cmdutils.Error(args, p.returncode, out, err)
2022-07-25T13:31:48.055712+0000 ost-basic-suite-master-host-0 vdsmd_init_common.sh[40924]: vdsm.common.cmdutils.Error: Command ['/usr/sbin/multipathd', 'show', 'config', 'local'] failed with rc=1 out=b'' err=b'Jul 25 13:31:48 | /etc/multipath.conf does not exist, blacklisting all devices.\nJul 25 13:31:48 | You can run "/sbin/mpathconf --enable" to create\nJul 25 13:31:48 | /etc/multipath.conf. See man mpathconf(8) for more details\n'
```

Looks like an issue with OST - trying to start vdsm *before* vdsm is configured (vdsm-tool configure --force), but it needs more investigation.

@aesteve-rh did you test fresh install on a new vm?

We may need to skip the multipath check if vdsm was not configured yet (e.g. /etc/multipath.conf not installed yet).

nirs commented on lib/vdsm/tool/configurators/multipath.py:
```diff
@@ -109,3 +105,24 @@ def isconfigured():
+    """
+    Check the multipath daemon configuration.
+    """
+    conf_metadata_ok = _check_mpath_metadata()
```
Checking multipath is possible only if conf_metadata_ok is "YES".

aesteve-rh commented:
/ost

nirs commented on lib/vdsm/tool/configurators/multipath.py:
```diff
@@ -109,3 +105,26 @@ def isconfigured():
+        if check_result.error is not None:
+            # Just warn for now, do not fail the isconfigured check here.
+            sys.stdout.write(
+                f"WARNING: Invalid configuration: {check_result.error}:\n")
```
If the check failed because running "multipathd show config local" failed, we log now:

    WARNING: Invalid configuration: error from multpathd...

Instead of:

    WARNING: Cannot check multipath configuration : error from multipathd...

aesteve-rh commented:
Now it fails with:

```
 2022-07-25T14:51:18.793734+0000 ost-basic-suite-master-host-1 vdsmd_init_common.sh[23057]:     raise ValueError(f"Invalid item: {item}")                                                                                                                                                                                     
 2022-07-25T14:51:18.794331+0000 ost-basic-suite-master-host-1 vdsmd_init_common.sh[23057]: ValueError: Invalid item: Section(name='blacklist', children=[Pair(key='devnode', value='^(rbd)[0-9]*'), Pair(key='devnode', value='!^(sd[a-z]|dasd[a-z]|nvme[0-9])'), Pair(key='protocol', value='(scsi:adt|scsi:sbp)'),          Section(name='device', children=[Pair(key='vendor', value='SGI'), Pair(key='product', value='Universal Xport')]), Section(name='device', children=[Pair(key='vendor', value='^DGC'), Pair(key='product', value='LUNZ')]), Section(name='device', children=[Pair(key='vendor', value='EMC'), Pair(key='product',               value='LUNZ')]), Section(name='device', children=[Pair(key='vendor', value='DELL'), Pair(key='product', value='Universal Xport')]), Section(name='device', children=[Pair(key='vendor', value='FUJITSU'), Pair(key='product', value='Universal Xport')]), Section(name='device', children=[Pair(key='vendor', value='IBM'),   Pair(key='product', value='Universal Xport')]), Section(name='device', children=[Pair(key='vendor', value='IBM'), Pair(key='product', value='S/390')]), Section(name='device', children=[Pair(key='vendor', value='LENOVO'), Pair(key='product', value='Universal Xport')]), Section(name='device',                           children=[Pair(key='vendor', value='(NETAPP|LSI|ENGENIO)'), Pair(key='product', value='Universal Xport')]), Section(name='device', children=[Pair(key='vendor', value='STK'), Pair(key='product', value='Universal Xport')]), Section(name='device', children=[Pair(key='vendor', value='SUN'), Pair(key='product',           value='Universal Xport')]), Section(name='device', children=[Pair(key='vendor', value='(Intel|INTEL)'), Pair(key='product', value='VTrak V-LUN')]), Section(name='device', children=[Pair(key='vendor', value='Promise'), Pair(key='product', value='VTrak V-LUN')]), Section(name='device', children=[Pair(key='vendor',     value='Promise'), Pair(key='product', value='Vess V-LUN')])]) 
```

Looks like the `multipath show conf` returned a weird output. I will check where this comes from. But it is mixing Pairs with Sections in the same level, and that is why it fails.

nirs commented:
Correct, the blacklist section support both pairs and sections. This is normal
and not related to OST.

We need to update tests to reflect real structure and change the parser to support
this.

We can skip the blackist section when checking since it cannot enable user friendly
names, but for parsing we need consume it.

aesteve-rh commented:
/ost

nirs commented on lib/vdsm/storage/mpathconf.py:
```diff
@@ -432,6 +435,58 @@ def read_metadata():
+    """
+    issues = []
+    for item in mpath_conf:
+        if not item.children or item.name == "blacklist":
```
Three is also blacklist_exceptions which can contain the same content as blacklist.

aesteve-rh replied:
True, I just checked the documentation. I have added both. The other sections are "defaults", "multipaths", and "devices", so we should be fine excluding those two.

nirs commented on tests/storage/mpathconf_test.py:
```diff
@@ -224,3 +224,133 @@ def test_mpathd_parser():
+            ]),
+            Section('name2', [
+                Pair('key', 'value')
+            ])
```
There is also blacklist_exceptions. We need to test both of them with the parser (currently it does not have mixed sections in the config), and in the checker.

nirs commented on lib/vdsm/tool/configurators/multipath.py:
```diff
@@ -109,3 +105,26 @@ def isconfigured():
+        if check_result.error is not None:
+            # Just warn for now, do not fail the isconfigured check here.
+            sys.stdout.write(
+                f"WARNING: {check_result.error}:\n")
```
It does not fit in one line now?

nirs commented on tests/storage/mpathconf_test.py:
```diff
@@ -172,3 +175,52 @@ def test_mpath_empty_conf(fake_mpath_conf):
+        key1 "value1"
+    }
+    entry {
+    }
```
We need to test here mixed section with both pairs and other sections (e.g. blacklist).

aesteve-rh commented:
/ost

aesteve-rh commented:
The new error looks unrelated:

```
2022-07-26T10:18:43.150594+0000 ost-basic-suite-master-host-0 vdsm[41758]: ERROR Exception raised
Traceback (most recent call last):                                                                                                                                                                                                                                                                                              
File "/usr/lib/python3.6/site-packages/vdsm/vdsmd.py", line 159, in run serve_clients(log)
File "/usr/lib/python3.6/site-packages/vdsm/vdsmd.py", line 120, in serve_clients  atexit._run_exitfuncs()                                                                                                                                                                                                                       
File "/usr/lib/python3.6/site-packages/vdsm/clientIF.py", line 392, in prepareForShutdown secret.clear()                                                                                                                                                                                                                                
File "/usr/lib/python3.6/site-packages/vdsm/virt/secret.py", line 96, in clear  for virsecret in con.listAllSecrets():                                                                                                                                                                                                        
File "/usr/lib/python3.6/site-packages/vdsm/common/libvirtconnection.py", line 131, in wrapper ret = f(*args, **kwargs)
File "/usr/lib/python3.6/site-packages/vdsm/common/function.py", line 94, in wrapper return func(inst, *args, **kwargs)                                                                                                                                                                                                                                                                                          File "/usr/lib64/python3.6/site-packages/libvirt.py", line 6164, in listAllSecrets raise libvirtError("virConnectListAllSecrets() failed")  
libvirt.libvirtError: invalid connection pointer in virConnectListAllSecrets 
```

aesteve-rh commented:
/ost
